README.md
# predictions

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-02-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-02-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/predictions/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/predictions/compare/0.1.0...0.1.1


src/predictions/test.clj
(ns ^{:author "Vladimir Urosevic"}
    predictions.test
    (:require [uncomplicate.neanderthal.core :refer :all]
             [uncomplicate.neanderthal.vect-math :refer :all]
             [uncomplicate.neanderthal.native :refer :all]))



(def x (dv [1.0 0.86 0.03 0.08 0.903 0.725 0.574 -0.1 -0.203 -0.031 -0.12413793 0.33 0.225]))
(def xx (dv [0 0.29	0.13 0.08	0.868	0.859	0.618	-0.208 -0.319 -0.125 -0.2	0.96 0.575]))
(def xxx (dv [1	1	1	1	0.918	0.87 0.615 -0.056	-0.144 0.047 -0.004344828	0.25 0.2]))

(defn random-number
  "random number in interval [0 .. 0.1]"
  [x]
  (rand 0.1))

(defn create-random-matrix
  "Initialize a layer"
  [dim-y dim-x]
  (dge dim-y dim-x (map random-number (replicate (* dim-x dim-y) 1))))

(defn layer-output
  [input weights o-func]
  (o-func (mv weights input)))

;; izmestiti (dv ...) van funkcije
(defn dtanh
  [y]
  (xpy (dv (replicate (dim y) 1)) (scal -1 (sqr (tanh y)))))

(defn output-deltas
  [targets outputs]
  (let [error (axpy targets (scal -1 outputs))]
    (mul (dtanh outputs) error)))

(defn hidden-error
  "output-deltas vector & copy weigths matrix & counter"
  [o-deltas weights current]
  (let [delta-scalar (entry o-deltas current)
        neuron-weights (row weights current)]
    (if (< current (dec (dim o-deltas)))
      (do
        (scal! delta-scalar neuron-weights)
        (hidden-error o-deltas weights (inc current)))
      (mv (trans weights) (dv (replicate (inc current) 1))))))

(defn hidden-deltas
  "hidden-error vector & hidden-output vector"
  [h-error h-output]
  (mul h-error (dtanh h-output)))

(defn change-output-weights
  [o-weights o-deltas o-hidden speed-learning current]
  (let [o-scalar (entry o-deltas current)
        neuron-weights (row o-weights current)]
      (do
        (axpy! (scal speed-learning (mul (dv (replicate (dim o-hidden) o-scalar)) o-hidden)) neuron-weights)
        (if (< (inc current) (dim o-deltas))
          (change-output-weights o-weights o-deltas o-hidden speed-learning (inc current)))
          )
      ;;(-> o-weights)
      ))

(defn change-hidden-weights
  [h-weights h-deltas i-hidden speed-learning current]
  (let [o-scalar (entry h-deltas current)
        neuron-weights (row h-weights current)]
      (do
        (axpy! (scal speed-learning (mul (dv (replicate (dim i-hidden) o-scalar)) i-hidden)) neuron-weights)
        (if (< (inc current) (dim h-deltas))
        (change-hidden-weights h-weights h-deltas i-hidden speed-learning (inc current)))
        )
      ;;(-> h-weights)
      ))

;; primer za vektor greske
(hidden-error (output-deltas (dv [0.93]) (dv [0.605])) (copy output-layer-test) 0)



;; hidden deltas, ideja...
(mul (mv output-layer-test (dv (replicate 24 1))) (output-deltas (dv [0.2 0.4 0.5]) (dv [0.1 0.12 0.43])))


(hidden-deltas
  (hidden-error (output-deltas (dv [0.913 0.762]) (dv [0.605 0.15])) (copy output-layer-test) 0)
  (layer-output x hidden-layer-test tanh)
  )

(change-hidden-weights  hidden-layer-test
                        (hidden-deltas (hidden-error (output-deltas
                                                       (dv [0.913 0.762])
                                                       (dv [0.914977105814607 0.7762420531257551])) (copy output-layer-test) 0) (layer-output x hidden-layer-test tanh))
                        x
                        0.7
                        0)

(change-output-weights output-layer-test
                       (output-deltas (dv [0.913 0.762]) (dv [0.914977105814607 0.7762420531257551]))
                       (layer-output x hidden-layer-test tanh)
                       0.7
                       0)


(defn backpropagation
  [hidden-layer output-layer input target speed-learning]
  (let [output (layer-output (layer-output input hidden-layer tanh) output-layer tanh)
        o-deltas (output-deltas target output)
        h-output (layer-output input hidden-layer tanh)
        h-deltas (hidden-deltas (hidden-error o-deltas (copy output-layer) 0) h-output)
        ]
    (do
      (change-hidden-weights hidden-layer h-deltas input speed-learning 0)
      (change-output-weights output-layer o-deltas h-output speed-learning 0)
      )))



(def hidden-layer-test (create-random-matrix 24 13))

(def output-layer-test (create-random-matrix 2 24))

(-> hidden-layer-test)


(layer-output x hidden-layer-test tanh)


(layer-output (layer-output x hidden-layer-test tanh) output-layer-test tanh)




test/predictions/core_test.clj
(ns ^{:author "Vladimir Urosevic"}
    predictions.core-test
    (:require [clojure.test :refer :all]
            [predictions.core :refer :all]
            [predictions.neuralnetwork :refer :all]
            [predictions.data :refer :all]
            [uncomplicate.neanderthal.core :refer :all]
            [uncomplicate.neanderthal.vect-math :refer :all]
            [uncomplicate.neanderthal.native :refer :all]
            [midje.sweet :refer [facts throws => roughly]]))



(facts
  "test unit vector"
  (prepare-unit-vector 1) => (dv [1])
  (prepare-unit-vector 5) => (dv [1 1 1 1 1])
  (prepare-unit-vector 0) => (dv [])
  )

(facts
  "test random number"
  (<= (random-number 0) 0.1) => true
  )

(facts
  "test create random matrix"
  (dim (create-random-matrix 5 3)) => 15
  )

(facts
  "test output layer"
  (layer-output (dv [1 1 1]) (dge 3 3 [1 1 1 1 1 1 1 1 1]) tanh) => (dv [0.9950547536867305 0.9950547536867305 0.9950547536867305])
  )

(facts
  "test dtanh"
  (dtanh (dv [1])) => (dv [0])
  )

(facts
  "test output deltas"
  (output-deltas (dv [0.813]) (dv [0.813])) => (dv [0])
  )




project.clj
;;   Copyright (c) Vladimir Urosevic. All rights reserved.
;;   The use and distribution terms for this software are covered by the
;;   Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php) or later
;;   which can be found in the file LICENSE at the root of this distribution.
;;   By using this software in any fashion, you are agreeing to be bound by
;;   the terms of this license.
;;   You must not remove this notice, or any other, from this software.

(defproject predictions "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.9.0"]
                 [org.slf4j/slf4j-api "1.7.25"]
                 [uncomplicate/neanderthal "0.18.0"]
                 [criterium/criterium "0.4.3"]
                 [midje/midje "1.9.1"]])


src/predictions/data.clj
(ns ^{:author "Vladimir Urosevic"}
   predictions.data
   (:require [clojure.string :as string]
            [uncomplicate.neanderthal.core :refer :all]
            [uncomplicate.neanderthal.vect-math :refer :all]
            [uncomplicate.neanderthal.native :refer :all]))

(defn parse-float [s]
    (Float/parseFloat s)
    )

(defn read-data-training
  "Read the csv file, split out each line and then each number, parse the tokens and break up the numbers so that the last is the target and everything else is the feature vector."
  []
  (as-> (slurp "resources/data_trening.csv") d
        (string/split d #"\n")
        (map #(string/split % #",") d)
        (map #(map parse-float %) d)
        (map (fn [s] {:x  (vec (drop-last s))  :y (last s)}) d)))

(defn read-data-test
  "Read the csv file, split out each line and then each number, parse the tokens and break up the numbers so that the last is the target and everything else is the feature vector."
  []
  (as-> (slurp "resources/data_test.csv") d
        (string/split d #"\n")
        (map #(string/split % #",") d)
        (map #(map parse-float %) d)
        (map (fn [s] {:x  (vec (drop-last s))  :y (last s)}) d)))


(def input-data-training (vec (map dv (map :x (read-data-training)))))
(def target-data-training (vec (map dv (map :y (read-data-training)))))

(def input-data-test (vec (map dv (map :x (read-data-test)))))
(def target-data-test (vec (map dv (map :y (read-data-test)))))


src/predictions/neuralnetwork.clj
(ns ^{:author "Vladimir Urosevic"}
    predictions.neuralnetwork
    (:require [uncomplicate.neanderthal.core :refer :all]
            [uncomplicate.neanderthal.vect-math :refer :all]
            [uncomplicate.neanderthal.native :refer :all]))

(def max-dim 2048)

(def unit-vector (dv (replicate max-dim 1)))

(defn prepare-unit-vector
  "preparing unit vector for other calculations"
  [n]
  (if (<= n max-dim)
    (subvector unit-vector 0 n)
    (dv [0])))

(defn random-number
  "random number in interval [0 .. 0.1]"
  [x]
  (rand 0.1))

(defn create-random-matrix
  "Initialize a layer"
  [dim-y dim-x]
  (dge dim-y dim-x (map random-number (replicate (* dim-x dim-y) 1))))

(defn layer-output
  [input weights o-func]
  (o-func (mv weights input)))

(defn dtanh
  [y]
  (xpy (prepare-unit-vector (dim y)) (scal -1 (sqr y))))

(defn output-deltas
  [targets outputs]
  (let [error (axpy targets (scal -1 outputs))]
    (mul (dtanh outputs) error)))

(defn hidden-error
  "output-deltas vector & copy weigths matrix & counter"
  [o-deltas weights current]
  (let [delta-scalar (entry o-deltas current)
        neuron-weights (row weights current)]
      (do
        (scal! delta-scalar neuron-weights)
        (if (< current (dec (dim o-deltas)))
        (hidden-error o-deltas weights (inc current))))
      (mv (trans weights) (prepare-unit-vector (inc current)))
    ))

(defn hidden-deltas
  "hidden-error vector & hidden-output vector"
  [h-error h-output]
  (mul h-error (dtanh h-output)))

(defn change-output-weights
  [o-weights o-deltas o-hidden speed-learning current]
  (let [o-scalar (entry o-deltas current)
        neuron-weights (row o-weights current)]
    (do
      (axpy! (scal speed-learning (mul (scal o-scalar (prepare-unit-vector (dim o-hidden))) o-hidden)) neuron-weights)
      (if (< (inc current) (dim o-deltas))
        (change-output-weights o-weights o-deltas o-hidden speed-learning (inc current))))))

(defn change-hidden-weights
  [h-weights h-deltas i-hidden speed-learning current]
  (let [o-scalar (entry h-deltas current)
        neuron-weights (row h-weights current)]
    (do
      (axpy! (scal speed-learning (mul (scal o-scalar (prepare-unit-vector (dim i-hidden))) i-hidden)) neuron-weights)
      (if (< (inc current) (dim h-deltas))
        (change-hidden-weights h-weights h-deltas i-hidden speed-learning (inc current))))))

(defn backpropagation
  [hidden-layer output-layer input target speed-learning]
  (let [output (layer-output (layer-output input hidden-layer tanh) output-layer tanh)
        o-deltas (output-deltas target output)
        h-output (layer-output input hidden-layer tanh)
        h-deltas (hidden-deltas (hidden-error o-deltas (copy output-layer) 0) h-output)]
    (do
      (change-hidden-weights hidden-layer h-deltas input speed-learning 0)
      (change-output-weights output-layer o-deltas h-output speed-learning 0))))

(defn learning-once
  [h-layer o-layer  input-vec target-vec speed-learning]
  (str (for [[i t] (map list input-vec target-vec)]
         ;; (for [a (replicate 1 1)]
         (backpropagation h-layer o-layer i t speed-learning)
         ;; )
         )))

(defn output-network
  [h-layer o-layer input-vec]
  (layer-output (layer-output input-vec h-layer tanh) o-layer tanh)
  )

(defn evaluation
  [h-layer o-layer input-vec target-vec]
  (for [[i t] (map list input-vec target-vec)]
    {:output (entry (output-network h-layer o-layer i) 0)
     :target (entry t 0)
     :percent-abs (Math/abs (* (/ (- (entry (output-network h-layer o-layer i) 0) (entry t 0)) (entry t 0)) 100))
     }
    ))

(defn evaluation_sum_abs
  "evaluation neural network - average report by absolute deviations"
  [h-layer o-layer input-vec target-vec]
  (let [u (count (map :percent-abs (evaluation h-layer o-layer input-vec target-vec)))
        s (reduce + (map :percent-abs (evaluation h-layer o-layer input-vec target-vec)))
        ]
    (/ s u)))

src/predictions/example.clj
(ns ^{:author "Vladimir Urosevic"}
     predictions.example
     (:require [predictions.neuralnetwork :refer :all]
            [predictions.data :refer :all]
            [uncomplicate.neanderthal.core :refer :all]
            [uncomplicate.neanderthal.vect-math :refer :all]
            [uncomplicate.neanderthal.native :refer :all]
            [criterium.core :refer :all]))

(def x (dv [1.0 0.86 0.03 0.08 0.903 0.725 0.574 -0.1 -0.203 -0.031 -0.12413793 0.33 0.225]))



(def hidden-layer-test (create-random-matrix 128 50))
(def output-layer-test (create-random-matrix 1 128))

(-> hidden-layer-test)
(-> output-layer-test)

(layer-output x hidden-layer-test tanh)

(layer-output (layer-output x hidden-layer-test tanh) output-layer-test tanh)

(entry (layer-output (layer-output x hidden-layer-test tanh) output-layer-test tanh) 0)

(entry (layer-output (layer-output x hidden-layer-test tanh) output-layer-test tanh) 1)

(predictions.neuralnetwork/backpropagation hidden-layer-test output-layer-test x (dv [0.213 0.715]) 0.5)

(str (for [a (replicate 100 1)]
       (predictions.neuralnetwork/backpropagation hidden-layer-test output-layer-test x (dv [0.715]) 0.05)
       ))


(str (for [a (replicate 10000 1)]
       (learning-once hidden-layer-test output-layer-test input-data-training target-data-training 0.005)
       ))


(evaluation hidden-layer-test output-layer-test input-data-training target-data-training)

(evaluation hidden-layer-test output-layer-test input-data-test target-data-test)
(evaluation_sum_abs hidden-layer-test output-layer-test input-data-test target-data-test)